### PR TITLE
Introduce `UnsafeBeanProperty` and fix performance of setter

### PR DESCRIPTION
**Pull request by dstepanov**

### CODE REVIEW

**Comment by graemerocher** [comment]
Looks good, if you fixed the tests etc.

It will probably need to target 3.4.x as well I imagine

**Comment by dstepanov** [comment]
I think it can be a part of 3.3 as there is only an experimental interface added. If not I can backport the perf. optimization.

**Review by graemerocher — APPROVED** [review]

File: core/src/main/java/io/micronaut/core/beans/UnsafeBeanProperty.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+     */
+    T getUnsafe(@NonNull B bean);
+
+    /**
+     * Unsafe version of {@link #withValue(Object, Object)}.
+     *
+     * @param bean  The bean
+     * @param value The new value
+     * @return Either the existing instance or the property is mutable or a newly created instance via the copy constructor pattern.
+     * @since 2.3.0
```
Comment (jameskleeh) [inline core/src/main/java/io/micronaut/core/beans/UnsafeBeanProperty.java]: since is incorrect

**Review by jameskleeh — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/micronaut/core/beans/AbstractBeanProperty.java [hunk 2]
```diff
@@ -152,6 +164,10 @@ public final void set(@NonNull B bean, @Nullable P value) {
         writeInternal(bean, value);
     }
 
+    @Override
+    public void setUnsafe(B bean, P value) {
```
Comment (graemerocher) [inline core/src/main/java/io/micronaut/core/beans/AbstractBeanProperty.java]: Can we make this `final`?

File: core/src/main/java/io/micronaut/core/beans/AbstractBeanProperty.java [hunk 3]
```diff
@@ -117,14 +119,24 @@ public final P get(@NonNull B bean) {
...
 
         if (!beanType.isInstance(bean)) {
             throw new IllegalArgumentException("Invalid bean [" + bean + "] for type: " + introspection.getBeanType());
         }
-        if (value == get(bean)) {
+        return withValueUnsafe(bean, value);
+    }
+
+    @Override
+    public B withValueUnsafe(B bean, P value) {
```
Comment (graemerocher) [inline core/src/main/java/io/micronaut/core/beans/AbstractBeanProperty.java]: Can we make this `final`?

File: core/src/main/java/io/micronaut/core/beans/AbstractBeanProperty.java [hunk 4]
```diff
@@ -117,14 +119,24 @@ public final P get(@NonNull B bean) {
         return readInternal(bean);
     }
 
+    @Override
+    public P getUnsafe(B bean) {
```
Comment (graemerocher) [inline core/src/main/java/io/micronaut/core/beans/AbstractBeanProperty.java]: Can we make this `final`?

**Review by graemerocher — CHANGES_REQUESTED** [review]

**Review by graemerocher — APPROVED** [review]

**Pull request merged**